tun/tun_illumos.go

```diff
@@ -0,0 +1,271 @@
+	 * "unix.OpenFile()" so that we get the os.File object here.  The rest
+	 * of the Wireguard API seems to depend on that functionality.
+	 */
+	tunFile, err := os.OpenFile(dev_node, unix.O_RDWR, 0)
+	if err != nil {
+		unix.Close(ip_fd)
+		return nil, fmt.Errorf("Could not open TUN (%s)", dev_node)
+	}
+	fd := int(tunFile.Fd())
+
+	/*
+	 * Ask the TUN driver for a new PPA number:
+	 */
+	var ppa int
+	var s unix.Strioctl
+	s.Cmd = unix.TUNNEWPPA
+	for try := 0; try < 128; try++ {
+		s.SetInt(try)
+		ppa, err = unix.IoctlSetStrioctlRetInt(fd, unix.I_STR, &s)
+		if err != nil {
+			if err == unix.EEXIST {
+				// PPA already in use, try the next one.
+				continue
+			}
+			unix.Close(ip_fd)
+			tunFile.Close()
+			return nil, err
+		}
+		break
+	}
+
+	name = fmt.Sprintf("tun%d", ppa)
+
+	/*
+	 * Open another temporary file descriptor to the TUN driver.
+	 * XXX It's not clear if this is actually needed, or if we could
+	 * reuse the fd we got above...
+	 */
+	if_fd, err := unix.Open(dev_node, unix.O_RDWR, 0)
+	if err != nil {
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, fmt.Errorf("Could not open second TUN (%s)",
+			dev_node)
+	}
+
+	/*
+	 * Push the IP module onto the new TUN device.
+	 */
+	if err = unix.IoctlSetString(if_fd, unix.I_PUSH, "ip"); err != nil {
+		unix.Close(if_fd)
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, err
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
+	// common x/sys/unix functions use uint. Hand-cast things to help
+	// the compiler figure it out.
+	req := int(unix.IF_UNITSEL)
+	if err = unix.IoctlSetPointerInt(if_fd, uint(req), ppa); err != nil {
+		unix.Close(if_fd)
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, err
+	}
+
+	muxid, err := unix.IoctlSetIntRetInt(ip_fd, unix.I_PLINK, if_fd)
+	if err != nil {
+		unix.Close(if_fd)
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, err
+	}
+
+	/*
+	 * XXX It would seem that we can now close this file descriptor,
+	 * because of the persistent link established above.
+	 */
+	unix.Close(if_fd)
+
+	var l unix.Lifreq
+	if err := l.SetName(name); err != nil {
+		unix.IoctlSetInt(ip_fd, unix.I_PUNLINK, muxid)
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, err
+	}
+
+	// illumos defines the ioctl number as a signed int, but the
+	// common x/sys/unix functions use uint. Hand-cast things to help
+	// the compiler figure it out.
+	reqnum := int(unix.SIOCSLIFMUXID)
+	l.SetLifruInt(muxid)
+	if err = unix.IoctlLifreq(ip_fd, uint(reqnum), &l); err != nil {
+		unix.IoctlSetInt(ip_fd, unix.I_PUNLINK, muxid)
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, err
+	}
+
+	tun := &NativeTun{
+		events:  make(chan Event, 10),
+		errors:  make(chan error, 1),
+		mtu:     mtu, /* XXX We should do something with the MTU! */
+		name:    name,
+		tunFile: tunFile,
+		ip_fd:   ip_fd,
+	}
+
+	defer func() {
+		/*
+		 * XXX For now, we'll just send a link up event straight away.
```

Comment:
Will need to be completed before merging.

Reply:
What is supposed to go in here?